server/base/src/main/java/org/apache/accumulo/server/metadata/RootGcCandidates.java

```diff
@@ -31,85 +29,73 @@
+    private final int version;
+
+    /*
+     * The root tablet will only have a single dir on each volume. Therefore, root file paths will
+     * have a small set of unique prefixes. The following map is structured to avoid storing the
+     * same dir prefix over and over in JSon and java.
+     *
+     * SortedMap<dir path, SortedSet<file name>>
+     */
+    private final SortedMap<String,SortedSet<String>> candidates;
+
+    public Data(int version, SortedMap<String,SortedSet<String>> candidates) {
+      this.version = version;
+      this.candidates = candidates;
+    }
   }
 
-  /*
-   * The root tablet will only have a single dir on each volume. Therefore root file paths will have
-   * a small set of unique prefixes. The following map is structured to avoid storing the same dir
-   * prefix over and over in JSon and java.
-   *
-   * SortedMap<dir path, SortedSet<file name>>
-   */
-  private SortedMap<String,SortedSet<String>> candidates;
-
   public RootGcCandidates() {
-    this.candidates = new TreeMap<>();
+    this.data = new Data(VERSION, new TreeMap<>());
   }
 
-  private RootGcCandidates(SortedMap<String,SortedSet<String>> candidates) {
-    this.candidates = candidates;
+  public RootGcCandidates(String jsonString) {
+    this.data = gson.fromJson(jsonString, Data.class);
+    checkArgument(data.version == VERSION, "Invalid Root Table GC Candidates JSON version %s",
+        data.version);
+    data.candidates.forEach((parent, files) -> {
+      checkArgument(!parent.isBlank(), "Blank parent dir in %s", data.candidates);
+      checkArgument(!files.isEmpty(), "Empty files for dir %s", parent);
+    });
   }
 
-  public void add(Iterator<StoredTabletFile> refs) {
-    refs.forEachRemaining(ref -> {
-      String parent = ref.getPath().getParent().toString();
-      candidates.computeIfAbsent(parent, k -> new TreeSet<>()).add(ref.getFileName());
-    });
+  public void add(Stream<StoredTabletFile> refs) {
+    refs.forEach(ref -> data.candidates
+        .computeIfAbsent(ref.getPath().getParent().toString(), k -> new TreeSet<>())
+        .add(ref.getFileName()));
   }
 
-  public void remove(Collection<String> refs) {
-    refs.forEach(ref -> {
-      Path path = new Path(ref);
-      String parent = path.getParent().toString();
-      String name = path.getName();
-
-      SortedSet<String> names = candidates.get(parent);
-      if (names != null) {
-        names.remove(name);
-        if (names.isEmpty()) {
-          candidates.remove(parent);
-        }
-      }
-    });
+  public void remove(Stream<String> refs) {
+    refs.map(Path::new).forEach(
+        path -> data.candidates.computeIfPresent(path.getParent().toString(), (key, values) -> {
+          values.remove(path.getName());
+          return values.isEmpty() ? null : values;
+        }));
   }
 
-  public Stream<String> stream() {
-    return candidates.entrySet().stream().flatMap(entry -> {
+  public Stream<String> sortedStream() {
```

Comment:
Why change this to a sorted stream?

Reply:
Because the caller was sorting, even though it was already sorted. I wanted to make it more clear in the method name and made sure that the method implementation returned a sorted stream.